### PR TITLE
markers: include folder name for marker files

### LINKED ISSUE #7155: problems: view should display folder name for a given file
**Description**

The `problems-view` should display the folder name for a given file when it contains markers.
At the moment, no folder name is present which can lead to confusion whenever a file with the same name is present. I believe this might be a regression since I vaguely remember it having the folder name in past releases.

image

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

Fixes #7155

The following pull-request adds the missing handler for `getLongName()` in the `MarkerTreeLabel` to correctly return the folder name when displaying files in the `problems-view`.

#### How to test

_Single Root Workspace:_

1. start the application and add errors/warnings to a number of files
2. open the `problems-view`, the folder name should be displayed alongside the file name

Screen Shot 2020-02-18 at 9 19 57 AM

_Multiple Root Workspace:_

1. start the application with a **multiple root workspace** and add errors/warnings to a number of files
2. open the `problems-view`, the folder name should be displayed alongside the file name along with the root name

Screen Shot 2020-02-18 at 9 19 20 AM

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
I've added some unit tests to be able to verify the behavior of `marker-tree-label-provider`.

**Comment by elaihau** [comment]
Tested on Gitpod.

Worked well in the single root workspace.

Worked in multi root workspace, and the change can be improved by adding the root folder name to the directory name, like what vscode does:

Screenshot from 2020-02-17 21-21-36

**Comment by vince-fugnitto** [comment]
> Worked in multi root workspace, and the change can be improved by adding the root folder name as part of the directory name, like what vscode does:

Thanks for the suggestion! I'll update the code shortly :)

**Comment by vince-fugnitto** [comment]
@elaihau @lmcbout I updated the code to handle multiple root workspaces as well.
In this scenario, the root name is appended after the directory (as seen in the screenshot).

File: packages/markers/src/browser/problem/problem-widget.tsx [hunk 1]
```diff
@@ -170,6 +174,25 @@ export class ProblemWidget extends TreeWidget {
...
+     * - In a multiple root workspace, the root name is appended after the directory.
+     * @param node the marker info node.
+     *
+     * @returns the `description` to be used for display purposes.
+     */
+    protected getDescription(node: MarkerInfoNode): string {
+        const description: string[] = [];
+        description.push(this.toNodeDescription(node));
+        if (this.workspaceService.isMultiRootWorkspaceOpened) {
+            const rootName = this.workspaceService.getWorkspaceRootUri(node.uri);
```
Comment (akosyakov) [inline packages/markers/src/browser/problem/problem-widget.tsx]: `rootUri`

File: packages/markers/src/browser/problem/problem-widget.tsx [hunk 2]
```diff
@@ -170,6 +174,25 @@ export class ProblemWidget extends TreeWidget {
...
     }
 
+    /**
+     * Get the directory name of the node to be used as a description.
+     * - In a multiple root workspace, the root name is appended after the directory.
+     * @param node the marker info node.
+     *
+     * @returns the `description` to be used for display purposes.
+     */
+    protected getDescription(node: MarkerInfoNode): string {
```
Comment (akosyakov) [inline packages/markers/src/browser/problem/problem-widget.tsx]: I wonder why it cannot be a part of `MarkerTreeLabelProvider.getLongName`?  It is the class responsible to compute such information.

Comment (vince-fugnitto) [reply packages/markers/src/browser/problem/problem-widget.tsx]: Good idea, I'll refactor and add a test for multiple root workspaces.

**Comment by vince-fugnitto** [comment]
Updated the code so that:
- `MarkerTreeLabelProvider` is solely responsible for creating the label, even in a multiple root workspace (`${directory} ● ${rootName}`)
- included additional test to handle the multiple root scenario

**Comment by lmcbout** [comment]
Tested  single root workspace with Chrome
- In probems view, only see the parent folder of the file, I don't see the whole uri from the root workspace, it is what I should expect!

> ex:  root/a/b/c/file1.txt   --> I see on problem view: file1.txt c (1)
> should it be: file1.txt root/a/b/c (1)

- Should squash the 2 commits into one before merging

**Comment by lmcbout** [comment]
If you have something like  "one error for each file"  in the following:
root/a/b/c/file1.txt
root/a/b/dd/file1.txt
root/a/b/file1.txt

It will be shown in problem view as
file1.txt c (1)
file1.txt dd (1)
file1.txt b (1)

 I think when a filename is found in different folder, may be more than just the last folderName should be visible on the problem view for a file. What do you think ?

**Comment by vince-fugnitto** [comment]
> Tested single root workspace with Chrome
> 
> * In probems view, only see the parent folder of the file, I don't see the whole uri from the root workspace, it is what I should expect?

It should be fine in the latest update.

> * Should squash the 2 commits into one before merging

I want to keep commits separate:
- the first commit is fixing the regression for the label, and adding multiple-root support.
- the other commit is adding missing tests for the `MarkerTreeLabelProvider` (this is independent of the first commit and the history is cleaner).

**Comment by lmcbout** [comment]
When I add a second root workspace and I modify a file under the second workspace, the whole folder path is shown. If a modify  (.theia/setting.json) on the second project , the folderpath is shown, but If I modify the file (.theia/setting.json) for the first project, the the folderpath only show (.theia), not the complete folder path as I see for the second project

**Comment by vince-fugnitto** [comment]
> When I add a second root workspace and I modify a file under the second workspace, the whole folder path is shown. If a modify (.theia/setting.json) on the second project , the folderpath is shown, but If I modify the file (.theia/setting.json) for the first project, the the folderpath only show (.theia), not the complete folder path as I see for the second project

@lmcbout can you perhaps share a screenshot, and steps to reproduce? 
The current format is a bit hard to follow.

**Comment by lmcbout** [comment]
Look at the small video, you have in the problem view some files (setting.json) under the initial project which has nothing eles than .theia andanother file (settings.json) under the second root project (test) which you see the complete folderpath where to find the file compare to the initial root project which only has the relative path only, not the absolute path.
Isue7157

**Comment by vince-fugnitto** [comment]
> Look at the small video, you have in the problem view some files (setting.json) under the initial project which has nothing eles than .theia andanother file (settings.json) under the second root project (test) which you see the complete folderpath where to find the file compare to the initial root project which only has the relative path only, not the absolute path.
> Isue7157

Your second `setting.json`, is it your user preferences? If so, then it is outside of the current roots.

**Comment by lmcbout** [comment]
THere is issues in user pref, workspace pref , so the path is still missing for the workspace preferences for the initial root project, but the path is available for the second root project

**Comment by vince-fugnitto** [comment]
> THere is issues in user pref, workspace pref , so the path is still missing for the workspace preferences for the initial root project, but the path is available for the second root project

The user preferences are common and does not exist in either root directory that is why it does not show the root label as part of the description.

**Comment by lmcbout** [comment]
I know you don't show the root path for the user preferences, what about the root path for the settings.json under theia for the root project ! 
In this case here, it should be /home/lmcbout/theia/theia/.theia 
and here the path is : .theia

**Comment by vince-fugnitto** [comment]
> I know you don't show the root path for the user preferences, what about the root path for the settings.json under theia for the root project !
> In this case here, it should be /home/lmcbout/theia/theia/.theia
> and here the path is : .theia

Based on my latest screenshot, this is the case.
The `settings.json` shows `.theia ● theia`

**Comment by lmcbout** [comment]
>Based on my latest screenshot, this is the case.
>The settings.json shows .theia ● theia

and it should be **/home/lmcbout/theia/theia/.theia**  not just **.theia**

**Comment by vince-fugnitto** [comment]
> > Based on my latest screenshot, this is the case.
> > The settings.json shows .theia ● theia
> 
> and it should be **/home/lmcbout/theia/theia/.theia** not just **.theia**

Why? `.theia` exists one level from the the first root's workspace, and `theia` identifies which root.

**Comment by lmcbout** [comment]
>>Based on my latest screenshot, this is the case.
>>The settings.json shows .theia ● theia

>>and it should be **/home/lmcbout/theia/theia/.theia** not just .theia

>Why? .theia exists one level from the the first root's workspace, and theia identifies which root.

Why do you show the absolute path for the other root folder for the file: settings.json 
i.e. : /home/lmcbout/theia/workspace/test/.theia

**Comment by vince-fugnitto** [comment]
> > > Based on my latest screenshot, this is the case.
> > > The settings.json shows .theia ● theia
> 
> > > and it should be **/home/lmcbout/theia/theia/.theia** not just .theia
> 
> > Why? .theia exists one level from the the first root's workspace, and theia identifies which root.
> 
> Why do you show the absolute path for the other root folder for the file: settings.json
> i.e. : /home/lmcbout/theia/workspace/test/.theia

Can I see your `.theia-workspace` file?

**Comment by lmcbout** [comment]
> Can I see your .theia-workspace file?

No because I create a workspace with no name

**Comment by vince-fugnitto** [comment]
> > Can I see your .theia-workspace file?
> 
> No because I create a workspace with no name

@lmcbout it exists under your home (`~/.theia/Untitled.theia-workspace`).

**Comment by vince-fugnitto** [comment]
@lmcbout I'm asking so I can reproduce the problem, on my end I don't see any issues.

**Comment by vince-fugnitto** [comment]
@lmcbout ping :)

**Comment by lmcbout** [comment]
@vince-fugnitto  I just rebase your PR and there are errors with the linter:
@theia/markers: /home/lmcbout/theia/theia/packages/markers/src/browser/marker-tree-label-provider.spec.ts
@theia/markers:   17:1  error  More than 1 blank line not allowed  no-multiple-empty-lines
@theia/markers: ✖ 1 problem (1 error, 0 warnings)

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto I just rebase your PR and there are errors with the linter:
> @theia/markers: /home/lmcbout/theia/theia/packages/markers/src/browser/marker-tree-label-provider.spec.ts
> @theia/markers: 17:1 error More than 1 blank line not allowed no-multiple-empty-lines
> @theia/markers: ✖ 1 problem (1 error, 0 warnings)

@lmcbout fixed :)
